File: src/org/apollo/game/model/combat/AttackAction.java
package org.apollo.game.model.combat;

import org.apollo.game.action.DistancedAction;
import org.apollo.game.model.Character;
import org.apollo.game.model.Character.InteractionMode;
import org.apollo.game.model.Item;
import org.apollo.game.model.NPC;
import org.apollo.game.model.Player;
import org.apollo.game.model.Position;
import org.apollo.game.model.Skill;
import org.apollo.game.model.World;
import org.apollo.game.scheduling.impl.BarrowsDegradingTask;

/**
 * The periodic attack event.
 * 
 * Credits to rs2-server team for formulas and a guideline
 * as how to structure it
 * @author The Wanderer
 */
@SuppressWarnings("rawtypes")
public class AttackAction extends DistancedAction {

    /**
     * Creates the attack event.
     * 
     * @param Character
     *            The entity that is attacking.
     * @param retaliating
     *            The retaliation flag.
     */
    @SuppressWarnings("unchecked")
    public AttackAction(Character character, boolean retaliating) {
        super(character.getCombatCooldownDelay(), !retaliating, character, character.getInteractingCharacter().getPosition(), character.getActiveCombatAction().distance(character.getInteractingCharacter()));
        System.out.println(character.getActiveCombatAction().distance(character.getInteractingCharacter()));
        System.out.println(character.getCombatState().getCombatStyle().getId());
        System.out.println(character.getCombatState().getCombatStyle());
    }

    @Override
    public QueuePolicy getQueuePolicy() {
        return QueuePolicy.FORCE;
    }

    @Override
    public WalkablePolicy getWalkablePolicy() {
        return WalkablePolicy.FOLLOW;
    }

    @SuppressWarnings("unused")
    @Override
    public void executeAction() {
        final Character character = getCharacter();
        InteractionMode mode = character.getInteractionMode();
        Character target = character.getInteractingCharacter();
        if (character.isDestroyed()) {
            if (target != null && target.getEventManager() == null) {
                ((NPC) target).setWalkOverridden(false);
            }
            this.stop(); // they disconnected/got removed
            return;
        }

        if (character.getEventManager() != null) {
            if (character.getSkillSet().getSkill(Skill.HITPOINTS).getCurrentLevel() <= 0) {
                stop();
                return;
            }
        }

        if (target == null || target.isDestroyed() || target.getCombatState().isDead() || mode != InteractionMode.ATTACK) {
            character.resetInteractingCharacter();
            if (character instanceof NPC) {
                ((NPC) character).setWalkOverridden(false);
            }
            if (target != null && target instanceof NPC) {
                ((NPC) target).setWalkOverridden(false);
            }
            target = null;
            this.stop(); // the target disconnected / got removed
            return;
        }

        final CombatAction action = character.getActiveCombatAction();

        if (!action.canHit(character, target)) {
            character.getCombatState().setQueuedSpell(null);
            character.resetInteractingCharacter();
            if (character instanceof NPC) {
                ((NPC) character).setWalkOverridden(false);
            }
            if (target instanceof NPC) {
                ((NPC) target).setWalkOverridden(false);
            }
            this.stop();
            return;
        }
        character.getWalkingQueue().clear();

        int extraDistance = 0;

        if ((character.getFirstDirection().toInteger() != -1 || character.getSecondDirection().toInteger() != -1)
                && (target.getFirstDirection().toInteger() != -1 || target.getSecondDirection().toInteger() != -1)) {
            extraDistance = 3;
        }
        int steps = Math.abs(character.getPosition().getY() - target.getPosition().getY());
        Position[] positions = new Position[steps + 1];
        positions[0] = target.getPosition();
        for(int i = 0; i < steps; i++) {
            positions[i + 1] = new Position(target.getPosition().getX(), target.getPosition().getY() + (steps + (i - steps)) + 1);
        }
        

        if(target.getEventManager() == null && target.isAutoRetaliating()) {
            for(Position pos : positions) {
                target.getWalkingQueue().addStep(pos);
            }
        }
        /*if (!character.getPosition().lineOfSight(target) || !character.getPosition().isWithinDistance(target.getPosition(), character.getActiveCombatAction().distance(character) + extraDistance)) {
        if (!character.getCombatState().isDead() && character.getCombatState().canMove()) {
        PathFinder pf = new RS2PathFinder();
        Path p = pf.findPath(character.getPosition(), character.getPosition().closestTileOfCharacter(target));
        
        
        if (p == null)
        return;
        
        if (p.getPoints().size() < 1)
        return;
        
        for (Point p2 : p.getPoints()) {
        character.getWalkingQueue().addStep(new Position(p2.getX(), p2.getY()));
        }
        character.getWalkingQueue().pulse();
        }
        this.setDelay(1);
        return;
        }*/

        this.setDelay(character.getCombatCooldownDelay());
        if (target != null && mode == InteractionMode.ATTACK) {
            action.hit(character, target);
            character.setInCombat(true);
            target.setInCombat(true);
            if (character.getEventManager() != null) {
                for (int i = 0; i < character.getEquipment().getItems().length; i++) {
                    for (int j = 0; j < BarrowsDegradingTask.barrowsNames.length; j++) {
                        if (character.getEquipment().get(i) != null &&
                                character.getEquipment().get(i).getDefinition().getName().contains(BarrowsDegradingTask.barrowsNames[j])) {
                            character.sendMessage("Setting barrows task in motion...");
                            Item item = character.getEquipment().get(i);
                            System.out.println(item.getTickInCombat());
                            World.getWorld().schedule(new BarrowsDegradingTask((Player) character, item));
                        }
                    }
                }
            }
            character.getCombatState().setAttackDelay(character.getCombatCooldownDelay());
        } else {
            this.stop(); // will be restarted later if another attack starts
        }
    }
}

File: data/plugins/combat/AttackAction.rb
require 'java'
java_import 'org.apollo.game.model.NPC'
java_import 'org.apollo.game.action.DistancedAction'
java_import 'org.apollo.game.model.Character'
java_import 'org.apollo.game.model.Character.InteractionMode'
java_import 'org.apollo.game.model.Animation'
java_import 'org.apollo.game.model.Graphic'
java_import 'org.apollo.game.model.Position'

class AttackAction < DistancedAction
  attr_reader :character, :retaliating
  
  def initialize(character, retaliating)
    super character.get_combat_cooldown_delay, !retaliating, character, character.get_interacting_character.get_position, character.get_active_combat_action.distance, character.get_interacting_character 
    @character = character
    @retaliating = retaliating
  end
  
  def execute_action
    character = get_character
    mode = character.get_interaction_mode
    target = character.get_interacting_character
    
    if character.is_destroyed
      iftarget != null && target.get_event_manager == null
      target.set_walk_overridden false
    end
    
    this.stop
    return
  end
    
  if target == null || target.is_destroyed || target.get_combat_state.is_dead || mode != InteractionMode.ATTACK
    character.reset_interacting_character
    
    if character.get_event_manager == null
      character.set_walk_overridden false
    end
    
    if target != null && target.get_event_manager == null
      target.set_walk_overridden false
    end
    
    target = null
    this.stop #the target disconnected / got removed
    return
  end

  action = character.get_active_combat_action

  if !action.can_hit character, target 
    character.get_combat_state.set_queued_spell null
    character.reset_interacting_character
    
    if character.get_event_manager == null
      character.set_walk_overridden false
    end
    
    if target.get_event_manager == null
      target.set_walk_overridden false
    end
    
    this.stop
    return
  end
  
  character.get_walking_queue.clear

  extra_distance = 0
    
  extra_distance = 3 if character.get_first_direction.to_integer != -1 || character.get_second_direction.to_integer != -1 && target.get_first_direction.to_integer != -1 || target.get_second_direction.to_integer != -1   

  #TODO: NPC Pathfinding. Then remove this.
  if !character.get_position.is_within_distance target.get_position, character.get_active_combat_action.distance(character.get_interacting_character) 
      this.stop
    end if character.get_event_manager == null
                    
  this.set_delay character.get_combat_cooldown_delay
  
  if target != null && mode == InteractionMode.ATTACK 
    action.hit character, target
    character.set_in_combat true
    target.set_in_combat true
  else 
    this.stop # will be restarted later if another attack starts
  end
end